tempesta_fw/http.c

```diff
@@ -259,27 +311,46 @@ tfw_http_send_resp(TfwHttpReq *req, TfwStr *msg, const TfwStr *date)
 		msg->len += crlf->len - crlf_len;
 	}
 
-	if (!(hmresp = tfw_http_msg_alloc_err_resp()))
-		return -ENOMEM;
+	if (!(hmresp = tfw_http_msg_alloc_err_resp())) {
+		TFW_DBG2("%s: Response message allocation error: conn=[%p]\n",
+			 __func__, req->conn);
+		goto err_create;
+	}
 	if (tfw_http_msg_setup(hmresp, &it, msg->len)) {
-		tfw_http_msg_free(hmresp);
-		return -ENOMEM;
+		TFW_DBG2("%s: Response skb allocation error: conn=[%p]\n",
+			 __func__, req->conn);
+		goto err_setup;
 	}
-
 	tfw_http_prep_date(date->ptr);
-	tfw_http_msg_write(&it, hmresp, msg);
-
+	if (tfw_http_msg_write(&it, hmresp, msg)) {
+		TFW_DBG2("%s: Respnse allocation error: conn=[%p]\n",
+			 __func__, req->conn);
+		goto err_setup;
+	}
 	tfw_http_resp_fwd(req, (TfwHttpResp *)hmresp);
-
-	return 0;
+	
+	return;
+err_setup:
+	tfw_http_msg_free(hmresp);
+err_create:
+	/*
+	 * Pairing of requests and responses gets broken here. As soon
+	 * as request is not linked with any response, sending to that
+	 * client stops starting with that request, because that creates
+	 * a "hole" in the chain of requests -- a request without a response.
+	 * Subsequent responses cannot be sent to the client until that
+	 * hole is closed, which at this point will never happen. To solve
+	 * this situation there is no choice but to close client connection.
+	 */
+	tfw_http_req_conn_close(req);
```

Comment:
The function name is too generic, better to rename to `tfw_http_resp_build_error()` or smth like that and move the comment to the function description.

Reply:
Corrected.